In [8]:
import warnings
warnings.filterwarnings("ignore")
import pandas as pd
import numpy as np

In [9]:
products_data=pd.read_csv('products_data.csv')
products_data

NameDescription  \
0      tungsten solder pick handle solder pick pickin...   
1      write right  screen protector sony tc screen n...   
2      casio men dbc databank  digital watch disconti...   
3      factoryreconditioned dewalt dwkr heavyduty  am...   
4      energizer   light twoway light feature bright ...   
...                                                  ...   
44799  berkley trilene xl  yard fishing line clear  l...   
44800  badger airbrush company basic spray gun set ea...   
44801  fimo soft polymer clay  ounces tropical green ...   
44802  wilton icing color  oz rose petal icing color ...   
44803  polyform sculpey iii polymer clay  oz french b...   

                      categories  
0          Arts, Crafts & Sewing  
1      Cell Phones & Accessories  
2      Clothing, Shoes & Jewelry  
3       Tools & Home Improvement  
4         Health & Personal Care  
...                          ...  
44799          Sports & Outdoors  
44800      Arts, Crafts & Sewing  
44801      Arts, Crafts & Sewing  
44802     Grocery & Gourmet Food  
44803      Arts, Crafts & Sewing  

[44804 rows x 2 columns]

# Rendre label numerique

In [10]:
l=[]
for text in products_data["categories"].unique():
    l.append(text)

y=[]
for text in products_data["categories"]:
    for j in range(21):
        if text==l[j]:
            y.append(j)
    
y = np.asarray(y)

yy = pd.Series(y)
yy

0         0
1         1
2         2
3         3
4         4
         ..
44799     9
44800     0
44801     0
44802    19
44803     0
Length: 44804, dtype: int64

# Extracting Features From TextData using TF-IDF

In [11]:
from sklearn import feature_extraction, model_selection, naive_bayes, pipeline, manifold, preprocessing

In [12]:
Text=[]
for text in products_data["NameDescription"]:
    Text.append(text)

In [13]:
from sklearn.feature_extraction.text import TfidfVectorizer
# First, we need a vectorizer object
vectorizer = TfidfVectorizer()
x = vectorizer.fit_transform(Text).toarray()
xx=  pd.DataFrame(x)
xx

0       1       2       3       4       5       6       7       8       \
0         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
1         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
3         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
44799     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
44800     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
44801     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
44802     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
44803     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

       9       ...  117187  117188  117189  117190  117191  117192  117193  \
0         0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
1         0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2         0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
3         0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4         0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
...       ...  ...     ...     ...     ...     ...     ...     ...     ...   
44799     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
44800     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
44801     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
44802     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
44803     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

       117194  117195  117196  
0         0.0     0.0     0.0  
1         0.0     0.0     0.0  
2         0.0     0.0     0.0  
3         0.0     0.0     0.0  
4         0.0     0.0     0.0  
...       ...     ...     ...  
44799     0.0     0.0     0.0  
44800     0.0     0.0     0.0  
44801     0.0     0.0     0.0  
44802     0.0     0.0     0.0  
44803     0.0     0.0     0.0  

[44804 rows x 117197 columns]

After feature extraction from our text it gives 117197 features.

# MLP befor feature selection

# 10 iteration

In [14]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,random_state=1,test_size=0.2)

In [15]:
#max_iter=500
from sklearn.neural_network import MLPClassifier
classifier = MLPClassifier(hidden_layer_sizes=(250,100,50), max_iter=10, warm_start=True, verbose=True)
classifier.fit(x_train,y_train)

Iteration 1, loss = 1.58689029
Iteration 2, loss = 0.34813436
Iteration 3, loss = 0.10393278
Iteration 4, loss = 0.04364380
Iteration 5, loss = 0.02622978
Iteration 6, loss = 0.02145862
Iteration 7, loss = 0.01895712
Iteration 8, loss = 0.01622078
Iteration 9, loss = 0.01522656
Iteration 10, loss = 0.01372657


MLPClassifier(hidden_layer_sizes=(250, 100, 50), max_iter=10, verbose=True,
              warm_start=True)

In [16]:
y_pred = classifier.predict(x_test)

In [17]:
from sklearn.metrics import accuracy_score
print("Number of features is", x_train.shape[1])
print("Test set accuracy of MLPClassifier Classifier is {}".format(round(accuracy_score(y_pred,y_test)*100,2)))

Number of features is 117197
Test set accuracy of MLPClassifier Classifier is 81.79


In [ ]:
# save model as pickle file

In [18]:
import pickle
pickle.dump(classifier, open('model(10Iteration).pkl', 'wb'))

In [21]:
loaded_model = pickle.load(open('model(10Iteration).pkl', 'rb'))
result = loaded_model.score(x_test, y_test)
print(result)

0.817877469032474
